### PR TITLE
fixes #20838 - adds redux based forms support to foreman

### PR DESCRIPTION
**Pull request by ohadlevy**
this pr adds basic client side validation forms using redux-form and redux-form-validations.

this specific PR implements the bookmark creation workflow.

### CODE REVIEW

**Comment by ohadlevy** [comment]
extracted storybook related changes into #4805 - once merged it will reduced the amount of change on this PR.

**Comment by ohadlevy** [comment]
@timogoebel I believe this PR is in ready to review/merge, would be interesting to see if you rebase #4597 easily on top?

**Comment by timogoebel** [comment]
@ohadlevy: Excellent. I'll test and review this tomorrow if time permits.

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: I don't think, this is the proper way to do this. Isn't there a way to do this without jQuery? How about passing this similar to the provider in the `MountingService`?

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: I can see developers wanting to overwrite this.
Shouldn't we use a placeholder and something similar to `sprintf` in the translated text instead of this concatenation?

**Review by timogoebel — COMMENTED** [review]
This looks pretty good as a first start into react based forms.
I have some minor comments, but I think we can get this in pretty fast to have a base for expansion. 👍

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: without jquery:
```js
document.querySelector('meta[name="csrf-token"]').content
```
i think passing it to the mount service is a lot of boilerplate code? TBH I'm not sure if its needed to begin with? i didnt re-read the auth code, but it might not even be need in some form?

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: also, passing it as a prop in mounting sevice feels like its boilerplate code, especially since its already on the dom.

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: sure, i can even see that not every form submit you would not want a toast, would you mind suggesting the alternative you want?

File: webpack/assets/javascripts/react_app/components/common/Alert/index.js
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/common/Alert/index.js:37]: Why is this needed?  Isn't this following http://www.patternfly.org/pattern-library/communication/toast-notifications/?

File: webpack/assets/javascripts/react_app/components/common/forms/Actions.js
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/common/forms/Actions.js]: One thing we ran into in katello when doing something very similar to this in angular was that after abstracting the form controls away like this we ended up needing several variations on them.  I wonder if we should make this configurable with sane defaults.

File: webpack/assets/javascripts/react_app/components/common/forms/Form.js
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/common/forms/Form.js:6]: Why is the form being put into a `.well`?

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (waldenraines) [reply webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: +1 for using the non-jquery method.  It would be great if we could eventually remove jQuery entirely 😁

**Review by waldenraines — COMMENTED** [review]
I still need to test this out, I was only able to look at the code this review.

File: webpack/assets/javascripts/react_app/components/common/forms/Form.js
Comment (timogoebel) [reply webpack/assets/javascripts/react_app/components/common/forms/Form.js:6]: ... because the ruby form helpers already do it that way.

https://github.com/theforeman/foreman/blob/16bd5715afbfdbea20f653435973908d395a8f7a/app/helpers/form_helper.rb#L284-L291

File: webpack/assets/javascripts/react_app/components/common/forms/Actions.js
Comment (tstrachota) [reply webpack/assets/javascripts/react_app/components/common/forms/Actions.js]: Large versatility could probably be easily achieved by adding simple `SubmitButton` and `CancelButton` components and provide them as default children for the Actions component.

**Comment by ohadlevy** [comment]
@tstrachota I dont have much capacity for this but would welcome the change, would you mind sending a PR to my branch?

File: webpack/assets/javascripts/react_app/components/common/Alert/index.js
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/common/Alert/index.js:37]: @waldenraines I suggest for us to start using patternfly-react and replace our components with theirs.

Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/common/Alert/index.js:37]: Agreed, see http://projects.theforeman.org/issues/21130.

File: webpack/assets/javascripts/react_app/components/common/forms/Form.js
Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/common/forms/Form.js:6]: I don't see patternfly using `.well` anywhere on their [forms](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/form.html).  Can we remove this from both ruby form helpers and from this implementation?

**Comment by ohadlevy** [comment]
@timogoebel What do you think on merging this PR? can we address the validation string extraction  (already a different redmine issue at http://projects.theforeman.org/issues/21134) in a follow up PR?

**Comment by ohadlevy** [comment]
rebased.

File: webpack/assets/javascripts/react_app/components/common/forms/Form.js
Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/common/forms/Form.js:6]: @ohadlevy thoughts on this?

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/common/forms/Form.js:6]: @waldenraines TBH its been like that every since we started with bootstrap, I don't have a strong opinion one way or another, but i don't think it should be done in this pr, if we want to change it, then it should be across the app and not only in this PR.

Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/common/forms/Form.js:6]: I'm okay with that, see http://projects.theforeman.org/issues/21171

File: webpack/assets/javascripts/react_app/components/common/forms/Actions.js
Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/common/forms/Actions.js]: Yeah, that's even better.  I'm fine doing that when needed to avoid premature optimization.

**Review by waldenraines — APPROVED** [review]
All my concerns have either been addressed or we have created new issues for them in redmine.  I am fine merging this.

**Comment by timogoebel** [comment]
> can we address the validation string extraction (already a different redmine issue at http://projects.theforeman.org/issues/21134) in a follow up PR?

@ohadlevy: I would have preferred to handle the client side validation in a separate PR. But I'm not against merging this. I don't have much capacity to do a final review now, but I'll try my best.

File: webpack/assets/javascripts/react_app/components/common/Alert/Alert.test.js
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/components/common/Alert/Alert.test.js]: why not make this a sentence?
`it('can include a title', ..`

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (timogoebel) [reply webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: I think, we can stick for this for now. But we should not concat the two stings in this way. Can we do it with a placeholder? `%s was successfully created.`
Also mind the dot after the sentence.

File: webpack/assets/javascripts/react_app/components/common/forms/Actions.js
Comment (timogoebel) [reply webpack/assets/javascripts/react_app/components/common/forms/Actions.js]: The `&nbsp` should go into the if condition below, i.e. only be added when the spinner is shown.
Looks a little off:
image

File: webpack/assets/javascripts/react_app/components/common/forms/Actions.js
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/components/common/forms/Actions.js:9]: I think, these buttons could use some space in between.

**Review by timogoebel — CHANGES_REQUESTED** [review]
@ohadlevy: Thanks. I think this looks good. I left some small comments inline.
I think, this is ready to merge. i18n is still missing, however we can do this in a separate PR. That would be a requirement for #4807, though.

**Comment by ohadlevy** [comment]
@timogoebel address all of your comments, please review again - thanks!

**Comment by ohadlevy** [comment]
@timogoebel tests are green :-)

File: webpack/assets/javascripts/react_app/components/common/Alert/Alert.test.js
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/components/common/Alert/Alert.test.js]: ... it can accept children ;-) But never mind.

**Review by timogoebel — APPROVED** [review]
... while I'm not a react expert, this looks like a pretty good start for me. 👍 and thanks, @ohadlevy.

**Comment by ohadlevy** [comment]
[test upgrade]

**Comment by ohadlevy** [comment]
@timogoebel its all yours :)

**Comment by timogoebel** [comment]
Merged, thanks @ohadlevy.
Please do rebase your other PR. :-)

**Pull request merged**
